gcc/rust/lex/rust-lex.h

```diff
@@ -38,13 +38,37 @@ struct RAIIFile
       fclose (file);
   }
 
+  static bool allowed_filetype (const struct stat &statbuf)
+  {
+    // The file could be either
+    // - a regular file
+    // - a char device (/dev/null...)
+    return S_ISREG (statbuf.st_mode) || S_ISCHR (statbuf.st_mode);
+  }
+
 public:
   RAIIFile (const char *filename) : filename (filename)
   {
     if (strcmp (filename, "-") == 0)
-      file = stdin;
+      {
+	file = stdin;
+      }
     else
-      file = fopen (filename, "r");
+      {
+	struct stat statbuf;
+	if (!(file = fopen (filename, "r")))
+	  {
+	    return;
+	  }
+
+	if (-1 == fstat (fileno (file), &statbuf)
+	    || !allowed_filetype (statbuf))
+	  {
+	    fclose (file);
+	    file = nullptr;
+	    errno = EISDIR;
```

Comment:
Setting errno here is IMO fine but we should maybe have a think about it. Should the RAIIFile class be responsible for emitting its own errors? As it is currently, we always check if `file_wrap.ok()` or `file_wrap.get_raw() != nullptr` before calling `rust_error_at` or `rust_fatal_error`. But I think it could be worth it to make this class a little more complex and self-contained to emit its own errors. With an optional flag to disable that behavior. But this can definitely be done later on as part of a cleanup and is fine for now :)